### DISCUSSION TITLE
Help with USB Direct

### DISCUSSION THREAD

**Discussion by clee604** [gtxaspec/wz_mini_hacks#212] [Q&A] [answered]
I'm looking to get some help getting a single cable USB Direct connection to work.

I've been trying to follow the guide here:
https://github.com/gtxaspec/wz_mini_hacks/blob/master/documentation/usb-direct.md

However I think I'm lacking the knowledge around the different types of networking configs (interfaces, dhcpcd.conf, systemd-networkd) with different distros of Linux. I'm using Linux Mint on a x64 and I see there's a /etc/network/interfaces but no dnsmasq.conf on the PC host. Are there there dependencies I need to install? When I turn on USB-Direct, the camera still connects by wifi and I see the usb0 is in a DOWN state.

Any help or guidance on what I should read up on is greatly appreciated.

**Comment by virmaior** [comment]
if the camera is still connecting with wifi, what's your wz_mini.conf look like, specifically are you using bonding or something like that?

here's my base config on my cameras:

```

#####NETWORKING#####
CUSTOM_HOSTNAME="WCV3Z1"

ENABLE_USB_ETH="false"
ENABLE_USB_ETH_MODULE_AUTODETECT="true"
ENABLE_USB_ETH_MODULE_MANUAL=""

ENABLE_USB_DIRECT="true"
USB_DIRECT_MAC_ADDR="02:FF:FF:FF:FF:01"

ENABLE_USB_RNDIS="false"

...

ENABLE_RTL8189FS_DRIVER="true"
ENABLE_ATBM603X_DRIVER="true"

##### NETWORK INTERFACE BONDING #####
BONDING_ENABLED="false"
BONDING_PRIMARY_INTERFACE="eth0"
BONDING_SECONDARY_INTERFACE="wlan0"
BONDING_LINK_MONITORING_FREQ_MS="100"                                           
BONDING_DOWN_DELAY_MS="5000"
BONDING_UP_DELAY_MS="5000"

```

basically the only variation is the   `USB_DIRECT_MAC_ADDR="02:FF:FF:FF:FF:01" `

with that config, the camera should be set for usb direct. 
If bonding is on, then if it fails to get a good connection it might be falling back to wifi.

just to emphasize because this was an immensely annoying problem **the usb mini to A cable that comes with the wyze camera does not support data, so it won't work.** (I had a bunch of other mini to A cables lying around and only 1 worked - so I bought a bunch).

Assuming the camera is correctly configured, then it should show up in `ifconfig` as a new interface called usb0. (I had one incident where I'd misconfigured and it was showing up as eth1).  

on my RPI,
I also get messages like when plugging in a camera in `tail /var/log/messages` or `dmesg`:
```
Jul 31 07:43:05 mulberrypi kernel: [1033891.956931] IPv6: ADDRCONF(NETDEV_CHANGE): usb0: link becomes ready
Jul 31 07:43:05 mulberrypi kernel: [1033891.957045] br0: port 1(usb0) entered blocking state
Jul 31 07:43:05 mulberrypi kernel: [1033891.957053] br0: port 1(usb0) entered forwarding state
Jul 31 07:43:40 mulberrypi kernel: [1033926.733938] usb 1-1.2.1.4: new high-speed USB device number 28 using xhci_hcd
Jul 31 07:43:40 mulberrypi kernel: [1033926.875041] usb 1-1.2.1.4: New USB device found, idVendor=0525, idProduct=a4a1, bcdDevice= 3.10
Jul 31 07:43:40 mulberrypi kernel: [1033926.875064] usb 1-1.2.1.4: New USB device strings: Mfr=1, Product=2, SerialNumber=0
Jul 31 07:43:40 mulberrypi kernel: [1033926.875082] usb 1-1.2.1.4: Product: NCM Gadget
Jul 31 07:43:40 mulberrypi kernel: [1033926.875099] usb 1-1.2.1.4: Manufacturer: wz_mini_ncm
Jul 31 07:43:40 mulberrypi kernel: [1033926.905907] cdc_ncm 1-1.2.1.4:1.0: MAC-Address: 02:b1:ef:a3:72:b2
Jul 31 07:43:40 mulberrypi kernel: [1033926.906937] cdc_ncm 1-1.2.1.4:1.0 usb5: register 'cdc_ncm' at usb-0000:01:00.0-1.2.1.4, CDC NCM, 02:b1:ef:a3:72:b2
```

Without doing any settings on your linux system, the ip address should be a random private assign ip. **That IP address is for the linux computer -- not the usb direct camera **, because each usb camera sets up a small network.

---
To recap:
1. please share the relevant parts of wz_mini.conf
2. for testing purposes, let's start without bonding
3. insure that your cable is a data cable
4. see if a new interface shows up in `ifconfig` when you plug it in (can also confirm with dmesg)

**Reply by clee604** [reply]
Thanks for the detailed reply, I spent some time trying to get it to work last night and made a bit of progress but to answer your questions:
1. I'm basically using the wz_conf as specified in your template guide:  usb_direct = true, bonding, eth, rdis false. I updated the MAC address as per your recommendation as well.
2. I was going to turn bonding (wifi) on after I got it working as you mentioned, it is currently off
3. This is what I finally made progress on last night, I also had to go through a drawer full of cables but I found one that has data now
4. Here is where I made progress after the data cable, but it also deviates from your guide:

The interface name for some reason is enx025a4bb25708 instead of usb0.

By default (no changes to any networking config) Linux Mint, the camera can't get any IP (internal or not). I'm using Linux Mint XFCE so there is a UI, there is a new 'Wired Connection 2' which corresponds to this new interface, if I update the network setting there to "Shared with other computers" then the cam gets an IP of 10.42.0.1 and the camera is accessible by the Wyze app. 

However it isn't accessible from the local network outside of the Linux Mint PC as it doesn't have an IP from the router. I'd like for the Wyze cam to get an IP from the router via DHCP on the same subnet (ie. 192.168.1.xxx). I'm assuming I need to set up some kind of bridge but here is where I got stuck. Alternatively if it's easier, I can set up a static IP, but I rather set the static IP at the router level if possible.

In short I'm able to get the camera to access internet now (Wyze Cam App works fine), but it doesn't pull an IP from the router (my DHCP server) so it isn't accessible via internal network (ie. RTSP via TinyCam).

**Comment by virmaior** [comment]
I.  In terms of USB direct then you've got the same sort of setup I do 
II. Trivia: from what I understand switching from usb0 to enx025a4bb25708  was something about creating unique network ids so as they don't collide in a series of attempts to sort out linux networking.
III. To access the usb direct subnet on the other computers, I think there are several possible solutions:

1. Figure out how to run the linux device as a bridge so that the usb direct cameras request their IPs from the router using DHCP (or static routing).
2. Add a route on the router itself.
3. Add routes on the computers you want to access this.
4. Don't and proxy things on the linux device (in your case Mint; in my case RPI)
5. Wireguard ?

So far I've been primarily working with 3. 
My cameras are USB-direct to a raspberry pi. Their subnet is 192.168.9.x/24. The network computers are on is 192.168.3.x/24 where the raspberry pi is 192.168.3.6.

By running,
```
sudo route -n add -net 192.168.9.0/24 192.168.3.6
``` 

on my mac, I can access all of the cameras because now they know to route over the RPI.

Solution 3 only works on a device-by-device basis though and non-rooted android doesn't provide access to route tables.

For 2, I've got it queued up for the next time the upstream router reboots, so hopefully that will give all of my devices access to the cameras directly.

For 1, I'm confident it is possible but I haven't figured it out yet. I think the basic setup would  be (1) turn off dhcp assigning that faces the usb-directs and (2) bridge the internet-facing interface with all of the usb-directs so that they all go to the main network and get their dhcp there.
I'm a bit worried to try this because my RPI is headless (no monitor, mouse, or keyboard of its own), so anything that messes up its networking will make it utterly inaccessible.

I've also done a bit for 4 (proxy) with all 5 cameras visible on a page web page on the RPI from any local device which updates the image every second -- it's not RTSP but it's something.

**Reply by clee604** [reply]
I realized I use rtsp-server to proxy the stream anyways on the Linux Mint device so I could use that to expose the RTSP stream. I wouldn't be able to SSH in the camera but I should be able to at least view the RTSP stream for the time being.

However for whatever reason I can't seem to access the 10.42.0.1 IP within my rtsp-server docker container. The docker container is already running in host mode so it should be able to access the internal IP no? Am I missing something here....

**Reply by clee604** [reply]
I can't seem to connect using VLC directly on the Linux Mint host so it's not the docker container. Something else is up... I can ping 10.42.0.1 but can't seem to access rtsp://user:pass@10.42.0.1:8554/unicast

**Reply by clee604** [reply]
New update... was able to RTSP and SSH from the host, but with IP address of 10.42.0.243.... i have no idea how the IP address is assigned since the wired connection says it is 10.42.0.1...

**Reply by clee604** [reply]
How did you set up your bonding if you are doing a static route? Wouldn't USB-Direct give you an internal (host only) IP and if it falls back to wifi you would get a network (router DHCP) IP? Or does your use case doesn't need static IPs?

**Reply by virmaior** [reply]
> New update... was able to RTSP and SSH from the host, but with IP address of 10.42.0.243.... i have no idea how the IP address is assigned since the wired connection says it is 10.42.0.1...

What do you mean the wired connection says it is 10.42.0.1?
my guess is that

-  10.42.0.1 is the IP address of the Mint Linux box on the USB direct network 
-  10.42.0.243 is the IP of the camera on USB direct network.

remember each usb direct create a small IP network.

> How did you set up your bonding if you are doing a static route? Wouldn't USB-Direct give you an internal (host only) IP and if it falls back to wifi you would get a network (router DHCP) IP? Or does your use case doesn't need static IPs?

dnsmasq.conf can provide static IPs to the bridged networks based on mac addresss

```
interface=br0
...
dhcp-range=br0,192.168.9.2,192.168.9.255,255.255.255.0,24h
dhcp-host=br0,02:FF:FF:FF:FF:01,192.168.9.101
dhcp-host=br0,02:FF:FF:FF:FF:02,192.168.9.102
dhcp-host=br0,02:FF:FF:FF:FF:03,192.168.9.103
dhcp-host=br0,02:FF:FF:FF:FF:04,192.168.9.104
dhcp-host=br0,02:FF:FF:FF:FF:05,192.168.9.105
dhcp-host=br0,02:FF:FF:FF:FF:07,192.168.9.107

```

this insures that they have those IPs

So in my case
- 192.168.9.1 is the IP of the RPI on that subnet
- 192.168.9.101 is wyze camera with MAC of 02:FF:FF:FF:FF:01
- 192.168.9.102 is wyze camera with MAC of 02:FF:FF:FF:FF:02
...

I have yet to switch to bonding  on the cameras -- mostly because this works for me and the fall back would be pretty similar -- the cameras were on a RPI-issued wifi network isolated from the internet.

**Reply by virmaior** [reply]
For docker, I'm guessing the issue was that you were using the wrong IP, but I'm not knowledgeable about that.

**Reply by clee604** [reply]
You're right i had figured out that 10.42.0.1 is the Mint box IP, and 10.42.0.234 is the Wyze cam IP. I only found out the wyze cam IP after the wyze cam app updated what its IP address was. Is there another way to determine it? ifconfig and ip a only show the linux box address and i can't ssh into camera without knowing the IP...

I'll have to try and figure out the static routes, but I'm trying to make the jump directly to the bridged adapter so it can pull an IP directly from the router which would bypass any need to use static routes at all.

**Reply by virmaior** [reply]
without dnsmasq , you can use `sudo nmap -F 10.42.0.0/24` on the ip range to find everything there. That would help finding the IP address the camera has.
It may also be possible to give serve it an IP in one of the configuration files `/etc/network/interfaces` or some other location.

**Reply by clee604** [reply]
is there a way to determine what network configuration my Linux Mint is using? I can see there is a /etc/network/interfaces but it's pretty much blank. There is also no dnsmasq.conf file I can find...

I can do some configuration in the UI of Mint, but I think its limited and I wasn't able to get the bridging to work there.

**Reply by virmaior** [reply]
unfortunately, I don't know. 

There's something like 4 ways of managing networks in linux:
1. /etc/network/interfaces
2. /etc/dhcpcd.conf
3. systemd's networkd
4. distro specific GUIs

dhcpd only incompletely supersedes interfaces so some things require going back to interfaces. networkd seems pretty robust. Most guides I found were for old school interfaces ... 

I think you'll need to it directly rather than in GUI merely because this is a bit of a custom setup.

**Comment by clee604** [comment, marked as answer]
I managed to get bridging to work, and with the GUI of all things.

My main mistake was not rebooting the camera, as the IP pull is on the camera side (even if I restart the network service on the linux host, the wyze camera still holds on to the old IP).

For those in the future looking to set this up, on the wyze camera just turn on usb-direct and connect to linux host. On the linux host, create a bridge between eth0 (or whatever your ethernet interface is to the local network) and the wyze camera network interface (could be usb0 or in my case was enx------). Make sure it set to Automatic DHCP if you want it pull an IP from router.

 In Linux mint GUI this duplicates the interfaces, so I had to remove the old 'wired connection' network (which generated a local 10.42.x network in my case). Reboot camera and I saw it pulled an IP via DHCP from my router.

Next step is to see if bonding with a wifi fallback will work, which i assume it should since now I have a local network IP, but I might have to set up the USB direct mac address to be the same as the wifi adapter? Will have to test... 

*edit* actually how would bonding work in this scenario? looking at the interfaces on the camera with usb-direct on, it looks like it already uses wlan0. Would i set up wlan0 as the primary, and wlanold as the secondary?

Many thanks your help @virmaior !

**Reply by clee604** [reply]
@gtxaspec I thought i had it fixed, but when I tried to do the same configuration it doesn't really survive reboots. I think what is happening is the network bridge isn't ready in time when the wyze cam is booting up, and falls back to wifi too soon. 

Is there a way to delay the wifi to test my theory? Or after the camera boots, is there a way to switch the connection to the usb-direct interface and pull a new IP via SSH?

**Reply by clee604** [reply]
edit2: i'm not actually sure what interface it uses when i connect using the usb0 assigned ip....

~~edit: nevermind it still connects through the wifi~~

~~manually bringing up usb0 and assigning an IP address using ip link/ifconfig and the bridge works, so there definitely is some timing issues going on while the camera is booting up with my configuration.~~

**Reply by gtxaspec** [reply]
@clee604 do you have bonding enabled?

**Reply by clee604** [reply]
no it is disabled, it seems with usb_direct enabled it falls back to wifi even if bonding is disabled

**Reply by gtxaspec** [reply]
and your setup is:

cam>laptop (or router?)>wlan/wired>main router?

**Reply by clee604** [reply]
set up like this:
wyze v3 usb -> linux mint host -> ethernet -> main router

I'm bridging the adapters on the linux mint host so the wyze cam pulls its own IP from the DHCP service from the router.

I just got it to work again but it seems totally random.. maybe like 1 out 20 reboots and it works fine. Other times the wyze falls back to wifi and I can see on the linux host the wyzecam network interface is not connected. If I manually raise usb0 on the wyzecam, the link is restored but the wyze is already setup to use wifi on wlan0

**Reply by gtxaspec** [reply]
what do the logs say? I would check dmesg, wz_user.sh

**Reply by clee604** [reply]
pulled the working log, let me reboot and see if i can get the non-working log as well

**Reply by clee604** [reply]
not working log:
```
+ first_run_check
+ '[' -e /opt/wz_mini/tmp/.wz_user_firstrun ]
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ echo 'wpa_supplicant.conf not ready, wait some time for creation.'
wpa_supplicant.conf not ready, wait some time for creation.
+ COUNT=0
+ ATTEMPTS=15
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 0 -eq 15 ]]
+ echo -e '0... \c'
0... + sleep 5
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ echo 'wpa_supplicant.conf not ready, wait some time for creation.'
wpa_supplicant.conf not ready, wait some time for creation.
+ COUNT=0
+ ATTEMPTS=15
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 0 -eq 15 ]]
+ echo -e '0... \c'
0... + sleep 5
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ wait_wlan
+ true
+ grep HWaddr
+ ifconfig wlan0
wlan0     Link encap:Ethernet  HWaddr D0:3F:27:20:2C:A0  
+ echo 'wlan0 hwaddr is up'
wlan0 hwaddr is up
+ store_mac
+ echo 'store original mac'
store original mac
+ + cat /sys/class/net/wlan0/address
tr '[:lower:]' '[:upper:]'
+ break
+ echo 'wpa_supplicant.conf ready'
wpa_supplicant.conf ready
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 1 -eq 15 ]]
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 1 -eq 15 ]]
+ '[' -f /opt/wz_mini/tmp/.T20 ]
+ KMOD_PATH=/opt/wz_mini/lib/modules/3.10.14__isvp_swan_1.0__
+ swap_enable
+ '[' -e /opt/wz_mini/swap ]
+ echo 'Swap file exists'
Swap file exists
+ grep mini
+ cat /proc/swaps
/opt/wz_mini/swap                       file		131068	0	-1
+ echo 'Swap is already enabled'
Swap is already enabled
+ '[[' false '==' true ]]
+ echo 'ipv6 disabled'
ipv6 disabled
+ sysctl -w 'net.ipv6.conf.all.disable_ipv6=1'
net.ipv6.conf.all.disable_ipv6 = 1
+ '[[' false '==' true ]]
+ echo nfsv4 disabled
nfsv4 disabled
+ '[[' false '==' true ]]
+ echo 'iptables disabled'
iptables disabled
+ '[[' false '==' true ]]
+ echo 'USB Ethernet disabled'
USB Ethernet disabled
+ '[[' true '==' true ]]
+ sed 's/^\(..\)\(..\)\(..\)\(..\)\(..\).*$/02:\1:\2:\3:\4:\5/'
+ md5sum
```

Working log
```
+ first_run_check
+ '[' -e /opt/wz_mini/tmp/.wz_user_firstrun ]
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ echo 'wpa_supplicant.conf not ready, wait some time for creation.'
wpa_supplicant.conf not ready, wait some time for creation.
+ COUNT=0
+ ATTEMPTS=15
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 0 -eq 15 ]]
+ echo -e '0... \c'
0... + sleep 5
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ echo 'wpa_supplicant.conf not ready, wait some time for creation.'
wpa_supplicant.conf not ready, wait some time for creation.
+ COUNT=0
+ ATTEMPTS=15
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 0 -eq 15 ]]
+ echo -e '0... \c'
0... + sleep 5
+ wpa_check
+ '[' -e /tmp/wpa_supplicant.conf ]
+ wait_wlan
+ true
+ grep HWaddr
+ ifconfig wlan0
wlan0     Link encap:Ethernet  HWaddr D0:3F:27:20:2C:A0  
+ echo 'wlan0 hwaddr is up'
wlan0 hwaddr is up
+ store_mac
+ echo 'store original mac'
store original mac
+ tr '[:lower:]' '[:upper:]'
+ cat /sys/class/net/wlan0/address
+ break
+ echo 'wpa_supplicant.conf ready'
wpa_supplicant.conf ready
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 1 -eq 15 ]]
+ '[[' -e /tmp/wpa_supplicant.conf ]]
+ '[[' 1 -eq 15 ]]
+ '[' -f /opt/wz_mini/tmp/.T20 ]
+ KMOD_PATH=/opt/wz_mini/lib/modules/3.10.14__isvp_swan_1.0__
+ swap_enable
+ '[' -e /opt/wz_mini/swap ]
+ echo 'Swap file exists'
Swap file exists
+ grep mini
+ cat /proc/swaps
/opt/wz_mini/swap                       file		131068	0	-1
+ echo 'Swap is already enabled'
Swap is already enabled
+ '[[' false '==' true ]]
+ echo 'ipv6 disabled'
ipv6 disabled
+ sysctl -w 'net.ipv6.conf.all.disable_ipv6=1'
net.ipv6.conf.all.disable_ipv6 = 1
+ '[[' false '==' true ]]
+ echo nfsv4 disabled
nfsv4 disabled
+ '[[' false '==' true ]]
+ echo 'iptables disabled'
iptables disabled
+ '[[' false '==' true ]]
+ echo 'USB Ethernet disabled'
USB Ethernet disabled
+ '[[' true '==' true ]]
+ sed 's/^\(..\)\(..\)\(..\)\(..\)\(..\).*$/02:\1:\2:\3:\4:\5/'
+ md5sum

```

**Reply by gtxaspec** [reply]
I don't see anything relating to USB direct, is there anything in the logs that mentions that? or g_ncm.ko?

**Reply by clee604** [reply]
the dmesg:
notworking
```
[   18.932261] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   19.891339] [atbm_log]:wlan0:free authen bss ++
[   19.934850] [atbm_log]:wlan0:free authen bss ++
[   20.638729] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   22.517846] using random self ethernet address
[   22.535849] using random host ethernet address
[   22.540983] using host ethernet address: 02:36:7f:47:01:b8
[   22.550331] using self ethernet address: 02:F0:FF:FF:FF:02<6>[   22.558737] usb0: HOST MAC 02:36:7f:47:01:b8
[   22.563160] dwc2 dwc2: ID PIN CHANGED!
[   22.569015] usb0: MAC 02:f0:ff:ff:ff:02
[   22.581251] g_ncm gadget: NCM Gadget
[   22.586571] g_ncm gadget: g_ncm ready
[   22.596999] dwc2 dwc2: session end detected!
[   23.010251] dwc2 dwc2: EP0 IN Time-out
[   23.015453] g_ncm gadget: high-speed config #1: CDC Ethernet (NCM)
[   23.061582] [atbm_log]:wlan0:free authen bss ++
[   23.104945] [atbm_log]:authen:(80:2a:a8:17:5f:ce),ssid(Nord(2))
[   23.130130] [atbm_log]:wlan0: authenticated
[   23.141913] [atbm_log]:wlan0:free authen bss ++
[   23.184947] [atbm_log]:wlan0:free authen bss --
[   23.198217] [atbm_log]:wlan0: associated
[   23.208083] [atbm_log]:[80:2a:a8:17:5f:ce]:20M channel
[   23.231966] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   23.687662] [atbm_log]:wlan0: deauthenticating from 80:2a:a8:17:5f:ce by local choice (reason=3)
[   23.755220] [atbm_log]:wlan0:free authen bss ++
[   23.760030] cfg80211: Calling CRDA to update world regulatory domain
[   23.795016] [atbm_log]:wlan0:free authen bss ++
[   23.938301] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   24.059803] [atbm_log]:wlan0:free authen bss ++
[   24.106915] [atbm_log]:authen:(44:d9:e7:f3:94:89),ssid(Nord(2))
[   24.160017] [atbm_log]:wlan0: authenticated
[   24.166726] [atbm_log]:wlan0:free authen bss ++
[   24.204887] [atbm_log]:wlan0:free authen bss --
[   24.217896] [atbm_log]:wlan0: associated
[   24.240512] [atbm_log]:[44:d9:e7:f3:94:89]:20M channel
[   24.250901] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   30.122098] <WIFI> rx ampdu ++ 
```

working
```
[   19.002261] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   19.961970] [atbm_log]:wlan0:free authen bss ++
[   20.005177] [atbm_log]:wlan0:free authen bss ++
[   20.729228] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   22.522950] using random self ethernet address
[   22.528509] using random host ethernet address
[   22.533153] using host ethernet address: 02:36:7f:47:01:b8
[   22.551852] using self ethernet address: 02:F0:FF:FF:FF:02<6>[   22.563205] dwc2 dwc2: ID PIN CHANGED!
[   22.567888] usb0: HOST MAC 02:36:7f:47:01:b8
[   22.574163] usb0: MAC 02:f0:ff:ff:ff:02
[   22.578357] g_ncm gadget: NCM Gadget
[   22.585649] g_ncm gadget: g_ncm ready
[   22.589511] dwc2 dwc2: session end detected!
[   23.027841] g_ncm gadget: high-speed config #1: CDC Ethernet (NCM)
[   23.171839] [atbm_log]:wlan0:free authen bss ++
[   23.204862] [atbm_log]:authen:(80:2a:a8:17:5f:ce),ssid(Nord(2))
[   23.250002] [atbm_log]:wlan0: authenticated
[   23.263650] [atbm_log]:wlan0:free authen bss ++
[   23.304836] [atbm_log]:wlan0:free authen bss --
[   23.320155] [atbm_log]:wlan0: associated
[   23.329544] [atbm_log]:[80:2a:a8:17:5f:ce]:20M channel
[   23.357254] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   23.677959] [atbm_log]:wlan0: deauthenticating from 80:2a:a8:17:5f:ce by local choice (reason=3)
[   23.755369] [atbm_log]:wlan0:free authen bss ++
[   23.760191] cfg80211: Calling CRDA to update world regulatory domain
[   23.794918] [atbm_log]:wlan0:free authen bss ++
```

**Reply by gtxaspec** [reply]
so the way the connection works, it will:

1. Bring up the usb0 interface
2. wait for wifi to be ready, connect to wifi
3. disconnect from wifi, rename interface wlan0 to wlan0old
4. rename usb0 to wlan0
5. bring up wlan0 interface
6. run dhcp client on the wlan0 interface, and wait for an IP from the usb interface.


once the wifi connection is down, it should not be up again...

**Reply by clee604** [reply]
From what I can tell, when it doesn't work it either doesn't bring up the usb0 interface or it doesn't keep it up for some reason. 

On the host it shows that it is disconnected
![image](https://user-images.githubusercontent.com/93049965/183268374-f97ffc84-aa7b-4bce-ad06-3eb230400f67.png)

When it works, or if I raise the usb0 up manually by SSH in the camera, it shows it is connected.
![image](https://user-images.githubusercontent.com/93049965/183268400-a64b2f3e-4fd9-4240-9c8d-74162313aeca.png)

Some kind of timing issue is the only thing I can think of right now because it would be working, and I would reboot the camera and it would go back to wifi again (and stay on wifi).

**Reply by clee604** [reply]
I suspect it has something to do with DCHP, because when I set a static network on the linux host it seems to connect more constantly. When raising the usb0 manually and setting an IP, I'm also not able to use `ifconfig usb0 dhcp start`, it gives an error of `ifconfig: bad address 'dhcp'` for some reason but manually specifying an ip/subnet and broadcast address works fine

**Reply by clee604** [reply]
@gtxaspec is there a way to set a static IP for the usb0 interface on the wyzecam side? Then I could eliminate DCHP as the variable.

**Reply by gtxaspec** [reply]
no, not at the moment.  can you test the new_init branch and see if that works better?

**Reply by clee604** [reply]
testing now, it got correctly got a DHCP ip on first boot, and it survived a reboot... so i think something in the init changes may have fixed it?  Is this going to be moved into the master branch or are you just testing something?

**Reply by gtxaspec** [reply]
yes, a few more changes and it will be merged to master.
